### PR TITLE
New Feature: Add 'callback': Function for style-formatter

### PR DESCRIPTION
**Pull request by SowmediaNL**
Hi,

To better control style-sheet generation i have created this new feature, where you can provide a callback to change the style-sheet before it is written.

(e.g. this is usefull for when you want to add a root style selector)
Documentation has been updated.

Tested and works fine.

### CODE REVIEW

File: lib/style-formatters/css.js [hunk 1]
```diff
@@ -12,7 +12,8 @@ module.exports = (symbols = [], options = {}) => {
         format: {
             type: 'data',
             publicPath: ''
-        }
+        },
+        callback: null
```
Comment (cascornelissen) [inline lib/style-formatters/css.js]: Let's default this to the following:

```
callback: (content) => content
```

File: lib/options-validator.js [hunk 2]
```diff
@@ -69,7 +69,11 @@ module.exports = (pattern, options = {}) => {
                         sizes: Joi.string(),
                         variables: Joi.string(),
                         mixin: Joi.string()
-                    })
+                    }),
+                    callback: Joi.alternatives([
+                        Joi.valid(null),
```
Comment (cascornelissen) [inline lib/options-validator.js]: I don't think `null` should be valid here, it should either be `undefined` or a function

File: lib/style-formatters/css.js [hunk 3]
```diff
@@ -46,8 +47,10 @@ module.exports = (symbols = [], options = {}) => {
         return `.${selector} { background-image: url("${content}"); }`;
     });
 
+    let contentOut = sprites.join('\n');
+
     return {
         warnings: [],
-        content: sprites.join('\n')
+        content: typeof options.callback === 'function' ? options.callback(contentOut) : contentOut
```
Comment (cascornelissen) [inline lib/style-formatters/css.js]: Then we can remove the `let` definition above and just do

```
content: options.callback(sprites.join('\n'))
```

File: docs/options.md [hunk 4]
```diff
@@ -57,7 +57,8 @@ new SVGSpritemapPlugin(string | string[], {
             sizes?: string,
             variables?: string,
             mixin?: string
-        }
+        },
+        callback?: function(content) | false,
```
Comment (cascornelissen) [inline docs/options.md]: I don't think `false` should be valid here

File: docs/options.md [hunk 5]
```diff
@@ -215,3 +216,10 @@ Name for the SCSS variable that is used for the Sass map containing [user-define
 
 #### `styles.variables.mixin` – `'sprite'`
 Name for the SCSS variable that is used for the Sass mixin.
+
+#### `styles.callback` – `false`
```
Comment (cascornelissen) [inline docs/options.md]: Can you change the default value to `undefined`?

File: lib/options-formatter.js [hunk 6]
```diff
@@ -103,7 +103,7 @@ module.exports = (pattern, options) => {
                 sizes: 'sizes',
                 variables: 'variables',
                 mixin: 'sprite'
-            }
+            },
```
Comment (cascornelissen) [inline lib/options-formatter.js]: Can you exclude this from the PR?

File: lib/style-formatters/less.js [hunk 7]
```diff
@@ -49,9 +50,11 @@ module.exports = (symbols = [], options = {}) => {
...
     });
 
+    let contentOut = fs.readFileSync(path.join(__dirname, '../templates/styles.less'), 'utf8')
+        .replace('/* SPRITES */', sprites.join('\n').trim());
+
     return {
         warnings: [],
-        content: fs.readFileSync(path.join(__dirname, '../templates/styles.less'), 'utf8')
-            .replace('/* SPRITES */', sprites.join('\n').trim())
+        content: typeof options.callback === 'function' ? options.callback(contentOut) : contentOut
```
Comment (cascornelissen) [inline lib/style-formatters/less.js]: Same comments as for the CSS formatter

File: lib/style-formatters/scss.js [hunk 8]
```diff
@@ -110,15 +111,17 @@ module.exports = (symbols = [], options = {}) => {
...
         warnings: output.warnings,
-        content: fs.readFileSync(path.join(__dirname, '../templates/styles.scss'), 'utf8')
-            .replace(/\/\* VAR_SPRITES \*\//g, options.variables.sprites.trim())
-            .replace(/\/\* VAR_SIZES \*\//g, options.variables.sizes.trim())
-            .replace(/\/\* VAR_VARIABLES \*\//g, options.variables.variables.trim())
-            .replace(/\/\* VAR_MIXIN \*\//g, options.variables.mixin.trim())
-            .replace(/\/\* SPRITES \*\//g, output.sprites.map((sprite) => `${indent()}${sprite}`).join(',\n').trim() || '/* EMPTY */')
-            .replace(/\/\* SIZES \*\//g, output.sizes.filter(Boolean).map((size) => `${indent()}${size}`).join(',\n').trim() || '/* EMPTY */')
-            .replace(/\/\* VARIABLES \*\//g, output.variables.filter(Boolean).map((variable) => `${indent()}${variable}`).join(',\n').trim() || '/* EMPTY */')
+        content: typeof options.callback === 'function' ? options.callback(contentOut) : contentOut
```
Comment (cascornelissen) [inline lib/style-formatters/scss.js]: Same comments as for the CSS formatter. I do like the separate variable definition in this case so let's do that for all of them but instead use `const content` and just use `content` in the return object.

**Comment by SowmediaNL** [comment]
Will this be accepted? Would be great! Thanks!

**Comment by cascornelissen** [comment]
@SowmediaNL, I don't see the changes, are you sure you pushed?

**Pull request closed without merging**